lib/src/git.rs

```diff
@@ -2213,17 +2343,22 @@ impl<'a> GitFetch<'a> {
     ) -> Result<IgnoredRefspecs, GitFetchError> {
         validate_remote_name(remote_name)?;
 
-        // check the remote exists
-        if self
+        // check the remote exists, but lazily read its configuration
+        let remote_result = self
             .git_repo
             .try_find_remote(remote_name.as_str())
-            .is_none()
-        {
-            return Err(GitFetchError::NoSuchRemote(remote_name.to_owned()));
-        }
+            .ok_or_else(|| GitFetchError::NoSuchRemote(remote_name.to_owned()))?
+            .map_err(|e| {
+                GitFetchError::InvalidRemoteConfiguration(remote_name.to_owned(), Box::new(e))
+            });
+
         // At this point, we are only updating Git's remote tracking branches, not the
         // local branches.
-        let expanded_refspecs = expand_fetch_refspecs(remote_name, branch_names)?;
+        let expanded_refspecs = if branch_names.is_empty() {
+            expand_fetch_refspecs_from_remote_definitions(remote_name, &remote_result?)
```

Comment:
I think the default fallback can be handled by caller. Then, we wouldn't have to add ignored refspecs, etc. to stats. It's also nicer that the library function doesn't have a special case for the empty case.

```rust
    // in commands/git/fetch.rs
    if branch_names.is_empty() {  // or .is_none()
        let refspecs = git::get_default_fetch_refspecs(remote_name)?;
        // warn if unsupported?
        refspecs
    } else {
        git::expand_fetch_refspecs(remote_name, branch_names)? // or wrap in enum?
    }
```

Maybe we can also extend the current `StringPattern` to support negative intersection, but that would require more work.